pkg/controller/registry/resolver/cache.go

```diff
@@ -0,0 +1,287 @@
+	return &OperatorCache{
+		logger:    logrus.New(),
+		rcp:       rcp,
+		snapshots: make(map[CatalogKey]*CatalogSnapshot),
+		ttl:       5 * time.Minute,
+		sem:       make(chan struct{}, 4),
+	}
+}
+
+type NamespacedOperatorCache struct {
+	snapshots map[CatalogKey]*CatalogSnapshot
+}
+
+var _ CatalogDependencyCache = &NamespacedOperatorCache{}
+
+func (c *OperatorCache) Namespaced(namespaces ...string) *NamespacedOperatorCache {
+	now := time.Now()
+	clients := c.rcp.ClientsForNamespaces(namespaces...)
+
+	result := NamespacedOperatorCache{
+		snapshots: make(map[CatalogKey]*CatalogSnapshot),
+	}
+
+	var misses []CatalogKey
+	c.m.RLock()
+	for key := range clients {
+		if snapshot, ok := c.snapshots[key]; ok && !snapshot.Expired(now) {
+			result.snapshots[key] = snapshot
+		} else {
+			misses = append(misses, key)
+		}
+	}
+	c.m.RUnlock()
+
+	if len(misses) == 0 {
+		return &result
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	// Take the opportunity to clear expired snapshots while holding the lock.
+	var expired []CatalogKey
+	for key, snapshot := range c.snapshots {
+		if snapshot.Expired(now) {
+			expired = append(expired, key)
+		}
+	}
+	for _, key := range expired {
+		delete(c.snapshots, key)
+	}
+
+	// Check for any snapshots that were populated while waiting to acquire the lock.
+	var found int
+	for i := range misses {
+		if snapshot, ok := c.snapshots[misses[i]]; ok && !snapshot.Expired(now) {
+			result.snapshots[misses[i]] = snapshot
+			misses[found], misses[i] = misses[i], misses[found]
+			found++
+		}
+	}
+	misses = misses[found:]
+
+	for _, miss := range misses {
+		s := CatalogSnapshot{
+			logger: c.logger.WithField("catalog", miss),
+			key:    miss,
+			expiry: now.Add(c.ttl),
+		}
+		s.m.Lock()
+		c.snapshots[miss] = &s
+		result.snapshots[miss] = &s
+		go c.populate(&s, clients[miss])
+	}
+
+	return &result
+}
+
+func (c *OperatorCache) populate(snapshot *CatalogSnapshot, provider ClientProvider) {
+	defer snapshot.m.Unlock()
+
+	c.sem <- struct{}{}
+	defer func() { <-c.sem }()
+
+	registry, err := provider.Get()
+	if err != nil {
+		snapshot.logger.Errorf("failed to connect to registry: %s", err.Error())
+		return
+	}
+
+	it, err := registry.ListBundles(context.TODO())
+	if err != nil {
+		snapshot.logger.Errorf("failed to list bundles: %s", err.Error())
+		return
+	}
+
+	var operators []Operator
+	for b := it.Next(); b != nil; b = it.Next() {
+		o, err := NewOperatorFromBundle(b, "", snapshot.key)
+		if err != nil {
+			snapshot.logger.Warnf("failed to construct operator from bundle, continuing: %s", err.Error())
+			continue
+		}
+		operators = append(operators, *o)
+	}
+	if err := it.Error(); err != nil {
+		snapshot.logger.Warnf("error encountered while listing bundles: %s", err.Error())
+	}
+
+	snapshot.operators = operators
+}
+
+type CatalogSnapshot struct {
+	logger    logrus.FieldLogger
+	key       CatalogKey
+	expiry    time.Time
+	operators []Operator
```

Comment:
I'm assuming this is just for simplicity now - we likely don't want to store the entire definition in the cache, just the subset required for resolution.

Reply:
Yes, I'll include that on the list of follow-up items to address once the resolver is integrated end-to-end.